incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -1026,49 +1106,95 @@ private void onKafkaData(
+                            .get((b, o, m) -> b.getStringWithoutLengthUtf8(o, m - o - keyPostfix.length()));
+                        final String16FW clientId = new String16FW(clientId0);
+
+                        if (type.equals(WILL_SIGNAL_NAME.asString()) && willDeliverIds.containsKey(clientId))
                         {
                             willDeliverIds.get(clientId).forEach(signaler::cancel);
-                            KafkaFetchWillProxy willFetcher = willFetchers.get(clientId);
+                            KafkaFetchWillStream willFetcher = willFetchers.get(clientId);
                             if (willFetcher != null)
                             {
                                 willFetcher.cleanup(traceId, authorization);
                             }
                         }
-                        break fetchWill;
+                        else if (type.equals(EXPIRY_SIGNAL_NAME.asString()) && sessionExpiryIds.containsKey(clientId))
+                        {
+                            signaler.cancel(sessionExpiryIds.get(clientId));
+                        }
+
+                        break reactToSignal;
                     }
-                    MqttWillSignalFW willSignal =
-                        mqttWillSignalRO.tryWrap(payload.buffer(), payload.offset(), payload.limit());
 
-                    if (willSignal != null)
-                    {
-                        long deliverAt = willSignal.deliverAt();
-                        final String16FW clientId = willSignal.clientId();
+                    final MqttSessionSignalFW sessionSignal =
+                        mqttSessionSignalRO.wrap(payload.buffer(), payload.offset(), payload.limit());
 
-                        if (deliverAt == MqttWillDeliverAt.UNKNOWN.value())
+                    if (sessionSignal != null)
+                    {
+                        switch (sessionSignal.kind())
                         {
-                            if (!instanceId.instanceId().equals(willSignal.instanceId()))
+                        case MqttSessionSignalFW.KIND_WILL:
+                            final MqttWillSignalFW willSignal = sessionSignal.will();
+                            long deliverAt = willSignal.deliverAt();
+                            final String16FW willClientId = willSignal.clientId();
+
+                            if (deliverAt == MqttTime.UNKNOWN.value())
                             {
-                                deliverAt = supplyTime.getAsLong() + willSignal.delay();
+                                if (!instanceId.instanceId().equals(willSignal.instanceId()))
+                                {
+                                    deliverAt = supplyTime.getAsLong() + willSignal.delay();
+                                }
+                                else
+                                {
+                                    break reactToSignal;
+                                }
```

Comment:
Probably we can simplify this with putting the break first (with inverted condition), then the `deliverAt = ...` doesn't need an `else` around it.

Actually, if you do that then the if conditions can be combined, so we don't need nested ifs either.